src/main/java/org/apache/sling/api/resource/FallbackValueMap.java

```diff
@@ -0,0 +1,216 @@
+ */
+
+package org.apache.sling.api.resource;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.apache.sling.api.wrappers.ValueMapDecorator;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public class FallbackValueMap implements ValueMap {
+    List<Resource> resources;
+
+    Set<Entry<String, Object>> entrySet;
+    Set<String> keySet;
+    Set<String> accessed = new HashSet<>();
+    boolean cacheFilled = false;
+    Collection<Object> values;
+    ValueMap cache = new ValueMapDecorator(new HashMap<>());
+
+    private static final String EMPTY = "";
+
+    private static final String IMMUTABLE_ERROR_MESSAGE = "Fallback value maps are immutable";
+
+    /**
+     * Main entry point for building the value map's list.
+     *
+     * @param resources var args resources from where value map will be computed, order matters
+     *                  here : will be the order of the lookup.
+     */
+    public void setResources(Resource... resources) {
+        this.resources = resources != null ? Arrays.asList(resources) : Collections.emptyList();
+    }
+
+    /**
+     * Main entry point for building the value map's list.
+     *
+     * @param resources list of resources from where value map will be computed, order matters
+     *                  here : will be the order of the lookup.
+     */
+    public void setResources(List<Resource> resources) {
+        this.resources = resources;
+    }
+
+    /**
+     * Constructor.
+     */
+    public FallbackValueMap() {}
+
+    public FallbackValueMap(Resource... resources) {
+        setResources(resources);
+    }
+
+    @Nullable
+    @Override
+    public <T> T get(@NotNull String s, Class<T> requiredClass) {
+        return getFromCache(s, requiredClass);
+    }
+
+    @Override
+    public <T> T get(@NotNull String s, T defaultValue) {
+        T value;
+        if (defaultValue == null) {
+            value = (T)get(s, String.class);
+        } else {
+            value = get(s, (Class<T>) defaultValue.getClass());
+        }
+        return value == null ? defaultValue : value;
+    }
+
+    @Override
+    public Object get(Object key) {
+        String k = key != null ? key.toString() : EMPTY;
+        return get(k, null);
+    }
+
+    /**
+     * Get value from cache (or fill the cache with it).
+     *
+     * @param s key
+     * @param <T> value's class
+     * @param type type requested
+     * @return value corresponding to the key. Note that if the key is "deep", that is
+     *          contains '/', an additional lookup will be done through the resources layers,
+     *          cache will be synced accordingly.
+     */
+    <T> T getFromCache(@NotNull String s, Class<T> type) {
```

Comment:
I would start with an implementation that doesn't cache. IMHO caching is a separate concern and could be yet another decorator. Also, imagine that an underlying ValueMap is modified - your cache would not reflect this. The user of the API needs to be able to opt in (or out), otherwise it will limit the utility of the class.